esp-hal-common/src/dma/mod.rs

```diff
@@ -67,11 +67,63 @@
 //! ));
 //! ```
 //!
-//! ⚠️ Note: Descriptors should be sized as `((CHUNK_SIZE + 4091) / 4092) * 3`.
-//! I.e., to transfer buffers of size `1..=4092`, you need 3 descriptors. The
-//! number of descriptors must be a multiple of 3.
+//! ⚠️ Note: Descriptors should be sized as `(CHUNK_SIZE + 4091) / 4092`.
+//! I.e., to transfer buffers of size `1..=4092`, you need 1 descriptor.
 
-use core::{marker::PhantomData, ptr::addr_of, sync::atomic::compiler_fence};
+use core::{marker::PhantomData, ptr::addr_of_mut, sync::atomic::compiler_fence};
+
+/// A DMA transfer descriptor.
+#[derive(Clone, Copy)]
+pub struct DmaDescriptor {
+    flags: u32,
+    buffer: *mut u8,
+    next: *mut DmaDescriptor,
+}
+
+impl DmaDescriptor {
+    /// An empty DMA descriptor used to initialize the descriptor list.
+    pub const EMPTY: Self = Self {
+        flags: 0,
+        buffer: core::ptr::null_mut(),
+        next: core::ptr::null_mut(),
+    };
+
+    fn set_size(&mut self, len: u16) {
+        let mask = 0b1111_1111_1111;
+        let bit_s = 0;
+        self.flags = (self.flags & !(mask << bit_s)) | (len as u32) << bit_s;
```

Comment:
This should probably mask the input `len & mask` and might as well make a helper function for this operation.